src/text-transform/filter.ts

```diff
@@ -0,0 +1,100 @@
+/* eslint-disable no-param-reassign */
+/* eslint-disable no-not-accumulator-reassign/no-not-accumulator-reassign */
+
+import evalExp from '@doc-tools/transform/lib/liquid/evaluation';
+
+import {ContentVariables} from './transformers';
+
+type WhenValue = string | boolean | undefined;
+type OrArray<T> = T | T[];
+type OrArrayRecord<T> = Record<string, OrArray<T>>;
+interface FilteredContentItem extends OrArrayRecord<FilteredContentItem> {}
+type FilteredContent = OrArray<FilteredContentItem>;
+type FilterableContentItem = FilteredContentItem & {when: WhenValue};
+export type FilterableContent = OrArray<FilterableContentItem>;
+
+function filterItems(
+    items: FilterableContentItem[],
+    vars: ContentVariables,
+    propertyName?: string,
+): FilterableContentItem[] | FilteredContentItem[] {
+    if (!Array.isArray(items)) {
+        throw new Error(
+            `Error while filtering: items has invalid key '${propertyName}' equals ${JSON.stringify(
+                items,
+            )}`,
+        );
+    }
+
+    return items.reduce<FilteredContentItem[]>((result, {when, ...item}) => {
+        const passedFiltration = checkWhenCondition(when, vars);
+
+        if (passedFiltration) {
+            const property = propertyName && item[propertyName];
+
+            if (property === undefined) {
+                result.push(item);
+            } else {
+                const filteredProperty = filterItems(
+                    property as FilterableContentItem[],
+                    vars,
+                    propertyName,
+                );
+
+                if (filteredProperty.length !== 0) {
+                    result.push({
+                        ...item,
+                        [propertyName as string]: filteredProperty,
+                    });
+                }
+            }
+        }
+
+        return result;
+    }, []);
+}
+
+function checkWhenCondition(whenValue: WhenValue, vars: ContentVariables) {
```

Comment:
its Function Declaration, but I think good practice is rearrange this func before `filterItems`

Reply:
It always was a controversial point where to place function dependencies before or after it (even in our team) and I think there is no strict rule for it